source/solvers/vof.cc

```diff
@@ -459,47 +481,393 @@ void
+                                           update_quadrature_points |
+                                           update_JxW_values |
+                                           update_gradients);
+
+  FEValues<dim> fe_values_pfg(*this->mapping,
+                              *this->fe_pfg,
+                              *this->cell_quadrature,
+                              update_values | update_quadrature_points |
+                                update_JxW_values | update_gradients);
+
+
+  // const unsigned int dofs_per_cell = this->fe->dofs_per_cell;
+  const unsigned int dofs_per_cell = this->fe_pfg->dofs_per_cell;
+
+  const unsigned int n_q_points = this->cell_quadrature->size();
+  FullMatrix<double> local_matrix_pfg(dofs_per_cell, dofs_per_cell);
+  Vector<double>     local_rhs_pfg(dofs_per_cell);
+  std::vector<types::global_dof_index> local_dof_indices(dofs_per_cell);
+
+  std::vector<Tensor<1, dim>> phi_pfg(dofs_per_cell);
+  std::vector<Tensor<2, dim>> phi_pfg_gradient(dofs_per_cell);
+
+  const FEValuesExtractors::Vector pfg(0);
+
+  std::vector<Tensor<1, dim>> phase_gradient_values(n_q_points);
+  std::vector<Tensor<1, dim>> pfg_values(n_q_points);
+  std::vector<Tensor<2, dim>> pfg_gradient_values(n_q_points);
+
+
+  system_rhs_pfg    = 0;
+  system_matrix_pfg = 0;
+
+  for (const auto &pfg_cell : this->pfg_dof_handler.active_cell_iterators())
+    {
+      if (pfg_cell->is_locally_owned())
+        {
+          typename DoFHandler<dim>::active_cell_iterator cell(
+            &(*this->triangulation),
+            pfg_cell->level(),
+            pfg_cell->index(),
+            &this->dof_handler);
+
+          fe_values_phase_fraction.reinit(cell);
+          fe_values_pfg.reinit(pfg_cell);
+
+          local_matrix_pfg = 0;
+          local_rhs_pfg    = 0;
+
+          fe_values_phase_fraction.get_function_gradients(
+            solution, phase_gradient_values);
+
+          fe_values_pfg[pfg].get_function_values(present_pfg_solution,
+                                                 pfg_values);
+
+          fe_values_pfg[pfg].get_function_gradients(present_pfg_solution,
+                                                    pfg_gradient_values);
+
+          for (unsigned int q = 0; q < n_q_points; ++q)
+            {
+              for (unsigned int k = 0; k < dofs_per_cell; ++k)
+                {
+                  phi_pfg[k]          = fe_values_pfg[pfg].value(k, q);
+                  phi_pfg_gradient[k] = fe_values_pfg[pfg].gradient(k, q);
+                }
+
+              for (unsigned int i = 0; i < dofs_per_cell; ++i)
+                {
+                  // Matrix assembly
+                  for (unsigned int j = 0; j < dofs_per_cell; ++j)
+                    {
+                      local_matrix_pfg(i, j) +=
+                        (phi_pfg[j] * phi_pfg[i] +
+                         simulation_parameters.surface_tension_force
+                             .phase_fraction_gradient_filter_value *
```

Comment:
I would put this constant in a temporary variable at the top instead of accessing it through the parameters all the time.

Reply:
I will apply this comment after adding the surface tension force, since now I don't have any idea about the values of these two filter constants and whether they should be equal.